trace_decoder/src/compact/compact_to_partial_trie.rs

```diff
@@ -8,183 +11,261 @@ use mpt_trie::{
+impl CompactToPartialTrieExtractionOutput for StorageTrieExtractionOutput {
+    fn process_code(&mut self, c_bytes: Vec<u8>) -> CompactParsingResult<()> {
+        Err(CompactParsingError::UnexpectedNodeForTrieType(
+            UnexpectedCompactNodeType::Code,
+            TrieType::Storage,
+        ))
+    }
 
-    Ok(())
+    fn process_leaf(
+        &mut self,
+        curr_key: Nibbles,
+        leaf_key: &Nibbles,
+        leaf_node_data: &LeafNodeData,
+    ) -> CompactParsingResult<()> {
+        process_leaf_common(
+            &mut self.trie,
+            curr_key,
+            leaf_key,
+            leaf_node_data,
+            |_, _| {
+                Err(CompactParsingError::UnexpectedNodeForTrieType(
+                    UnexpectedCompactNodeType::AccountLeaf,
+                    TrieType::Storage,
+                ))
+            },
+        )
+    }
+
+    fn get_trie(&mut self) -> &mut HashedPartialTrie {
+        &mut self.trie
+    }
 }
 
-fn process_leaf(
+fn process_leaf_common<F: FnMut(&AccountNodeData, &Nibbles) -> CompactParsingResult<Vec<u8>>>(
+    trie: &mut HashedPartialTrie,
     curr_key: Nibbles,
     leaf_key: &Nibbles,
     leaf_node_data: &LeafNodeData,
-    output: &mut CompactToPartialTrieExtractionOutput,
+    mut account_leaf_proc_f: F,
 ) -> CompactParsingResult<()> {
     let full_k = curr_key.merge_nibbles(leaf_key);
 
     let l_val = match leaf_node_data {
         LeafNodeData::Value(v_bytes) => rlp::encode(&v_bytes.0).to_vec(),
-        LeafNodeData::Account(acc_data) => {
-            convert_account_node_data_to_rlp_bytes_and_add_any_code_to_lookup(acc_data, output)
-        }
+        LeafNodeData::Account(acc_data) => account_leaf_proc_f(acc_data, &full_k)?,
     };
 
-    output.trie.insert(full_k, l_val);
+    trie.insert(full_k, l_val);
     Ok(())
 }
 
-fn process_extension(
+pub(super) fn create_partial_trie_from_remaining_witness_elem(
+    remaining_entry: WitnessEntry,
+) -> CompactParsingResult<StateTrieExtractionOutput> {
+    let remaining_node = remaining_entry
+        .into_node()
+        .expect("Final node in compact entries was not a node! This is a bug!");
+
+    create_partial_trie_from_compact_node(remaining_node)
+}
+
+pub(super) fn create_storage_partial_trie_from_compact_node(
+    node: NodeEntry,
+) -> CompactParsingResult<StorageTrieExtractionOutput> {
+    create_partial_trie_from_compact_node(node)
+}
+
+fn create_partial_trie_from_compact_node<T: CompactToPartialTrieExtractionOutput + Default>(
+    node: NodeEntry,
+) -> CompactParsingResult<T> {
+    let mut output = T::default();
+    create_partial_trie_from_compact_node_rec(Nibbles::default(), &node, &mut output)?;
+
+    Ok(output)
+}
+
+// TODO: Consider putting in some asserts that invalid nodes are not appearing
+// in the wrong trie type (eg. account )
+pub(super) fn create_partial_trie_from_compact_node_rec<
+    T: CompactToPartialTrieExtractionOutput + ?Sized,
+>(
     curr_key: Nibbles,
```

Comment:
Can this be private perhaps? Looks like it's only called from this file.
Also, I think it's more readable to use a `where` clause.

Reply:
Agree with both.